oak_functions/loader/src/metrics.rs

```diff
@@ -0,0 +1,283 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use rand::{distributions::Open01, thread_rng, Rng};
+use serde::Deserialize;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex},
+};
+
+/// Configuration for differentially-private metrics reporting.
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PrivateMetricsConfig {
+    /// The privacy budget. See
+    /// https://desfontain.es/privacy/differential-privacy-in-practice.html for more information
+    /// on epsilon-differential privacy and Laplacian noise.
+    pub epsilon: f64,
+    /// The number of requests that will be aggregated into each batch.
+    pub batch_size: usize,
+    /// The labels for events that can be reported. Any other events will be ignored.
+    pub allowed_labels: Vec<String>,
+}
+
+/// Aggregator for count-based differentially private metrics. The request metrics are released in
+/// aggregated batches after differentially private noise has been added. Once the number of
+/// requests reaches the batch threshold the metrics are logged and the request count and bucket
+/// counts are reset to 0.
+pub struct PrivateMetricsAggregator {
+    /// The request count.
+    count: usize,
+    /// The privacy budget.
+    epsilon: f64,
+    /// The number of requests that will be aggregated into each batch.
+    batch_size: usize,
+    /// The storage for event count buckets.
+    events: HashMap<String, usize>,
+}
+
+impl PrivateMetricsAggregator {
+    pub fn new(config: &PrivateMetricsConfig) -> anyhow::Result<Self> {
+        anyhow::ensure!(config.epsilon > 0.0, "Epsilon must be positive.");
+        Ok(Self {
+            count: 0,
+            epsilon: config.epsilon,
+            batch_size: config.batch_size,
+            events: config
+                .allowed_labels
+                .iter()
+                .map(|label| (label.clone(), 0))
+                .collect(),
+        })
+    }
+
+    /// Reports new events that should be included in the aggregated counts. If the number of
+    /// requests do not yet match the batch size `None` is returned. If the batch threshold
+    /// is reached the aggregated metrics for the batch are returned and the request count and
+    /// bucket counts are reset to 0. By default Laplacian noise is added to each of the aggregated
+    /// bucket counts that are returned. The metrics are returned as a tuple containing the batch
+    /// size and a vector of tuples cotaining the label and count for each bucket.
+    ///
+    /// If the `oak-unsafe` feature is enabled, or this function is called from a unit test, no
+    /// noise is added.
+    pub fn report_events(
+        &mut self,
+        events: HashSet<String>,
+    ) -> Option<(usize, Vec<(String, i64)>)> {
```

Comment:
(optional) alternatively you could return `HashMap<String, i64>` here too. Or did you consider and decide against it?